apps/precinct-scanner/src/screens/PollWorkerScreen.tsx

```diff
@@ -43,141 +51,241 @@ const PollWorkerScreen = ({
+        debug(
+          `Warning, ballot count from calculated tally (${tally.numberOfBallotsCounted}) does not match number of CVRs (${castVoteRecords.length}) `
+        )
+      }
+      setCurrentTally(tally)
     }
+    void calculateTally()
+  }, [])
+
+  const saveTally = async () => {
+    assert(currentTally)
+    const serializedTally = serializeTally(election, currentTally)
     await saveTallyToCard({
       tallyMachineType: TallySourceMachineType.PRECINCT_SCANNER,
-      totalBallotsScanned: castVoteRecords.length,
+      totalBallotsScanned: scannedBallotCount,
       isLiveMode,
       isPollsOpen: !isPollsOpen, // When we are saving we are about to either open or close polls and want the state to reflect what it will be after that is complete.
-      tally,
+      tally: serializedTally,
       metadata: [
         {
           machineId: machineConfig.machineId,
           timeSaved: Date.now(),
-          ballotCount: castVoteRecords.length,
+          ballotCount: scannedBallotCount,
         },
       ],
     })
   }
 
-  const { election } = electionDefinition
+  const printTallyReport = async () => {
+    await printer.print({ sides: 'one-sided' })
+  }
+
   const precinct = election.precincts.find((p) => p.id === currentPrecinctId)
 
   const [confirmOpenPolls, setConfirmOpenPolls] = useState(false)
   const openConfirmOpenPollsModal = () => setConfirmOpenPolls(true)
   const closeConfirmOpenPollsModal = () => setConfirmOpenPolls(false)
-  const openPollsAndSaveZeroReport = async () => {
-    setIsSavingTallyToCard(true)
-    await calculateAndSaveTally()
+  const openPollsAndHandleZeroReport = async () => {
+    setIsHandlingTallyReport(true)
+    if (hasPrinterAttached) {
+      await printTallyReport()
+    } else {
+      await saveTally()
+    }
     togglePollsOpen()
-    setIsSavingTallyToCard(false)
+    setIsHandlingTallyReport(false)
     closeConfirmOpenPollsModal()
   }
 
   const [confirmClosePolls, setConfirmClosePolls] = useState(false)
   const openConfirmClosePollsModal = () => setConfirmClosePolls(true)
   const closeConfirmClosePollsModal = () => setConfirmClosePolls(false)
-  const closePollsAndSaveTabulationReport = async () => {
-    setIsSavingTallyToCard(true)
-    await calculateAndSaveTally()
+  const closePollsAndHandleTabulationReport = async () => {
+    setIsHandlingTallyReport(true)
+    if (hasPrinterAttached) {
+      await printTallyReport()
+    } else {
+      await saveTally()
+    }
     togglePollsOpen()
-    setIsSavingTallyToCard(false)
+    setIsHandlingTallyReport(false)
     closeConfirmClosePollsModal()
   }
 
   const precinctName = precinct === undefined ? 'All Precincts' : precinct.name
+  const precinctSelection =
+    precinct === undefined
+      ? typedAs<PrecinctSelection>({ kind: PrecinctSelectionKind.AllPrecincts })
+      : typedAs<PrecinctSelection>({
+          kind: PrecinctSelectionKind.SinglePrecinct,
+          precinctId: precinct.id,
+        })
```

Comment:
There's no real need to use `typedAs` in this context. It's useful when you have an inline expression and cannot otherwise provide a type annotation. Here, however, you can easily add a type annotation:

```ts
  const precinctSelection: PrecinctSelection =
    precinct === undefined
      ? { kind: PrecinctSelectionKind.AllPrecincts }
      : {
          kind: PrecinctSelectionKind.SinglePrecinct,
          precinctId: precinct.id,
        }
```